crypto/ec/ecp_secp384r1.c

```diff
@@ -0,0 +1,15385 @@
+    uint64_t x3 = (uint64_t)(x1 & UINT64_C(0xffffffffffffffff));
+    *out1 = x3;
+    *out2 = (fiat_secp384r1_uint1)(0x0 - x2);
+}
+
+/*
+ * The function fiat_secp384r1_mulx_u64 is a multiplication, returning the full
+ * double-width result. Postconditions: out1 = (arg1 * arg2) mod 2^64 out2 =
+ * ⌊arg1 * arg2 / 2^64⌋
+ *
+ * Input Bounds:
+ *   arg1: [0x0 ~> 0xffffffffffffffff]
+ *   arg2: [0x0 ~> 0xffffffffffffffff]
+ * Output Bounds:
+ *   out1: [0x0 ~> 0xffffffffffffffff]
+ *   out2: [0x0 ~> 0xffffffffffffffff]
+ */
+static void fiat_secp384r1_mulx_u64(uint64_t *out1, uint64_t *out2,
+                                    uint64_t arg1, uint64_t arg2) {
+    fiat_secp384r1_uint128 x1 = ((fiat_secp384r1_uint128)arg1 * arg2);
+    uint64_t x2 = (uint64_t)(x1 & UINT64_C(0xffffffffffffffff));
+    uint64_t x3 = (uint64_t)(x1 >> 64);
+    *out1 = x2;
+    *out2 = x3;
+}
+
+/*
+ * The function fiat_secp384r1_cmovznz_u64 is a single-word conditional move.
+ * Postconditions:
+ *   out1 = (if arg1 = 0 then arg2 else arg3)
+ *
+ * Input Bounds:
+ *   arg1: [0x0 ~> 0x1]
+ *   arg2: [0x0 ~> 0xffffffffffffffff]
+ *   arg3: [0x0 ~> 0xffffffffffffffff]
+ * Output Bounds:
+ *   out1: [0x0 ~> 0xffffffffffffffff]
+ */
+static void fiat_secp384r1_cmovznz_u64(uint64_t *out1,
+                                       fiat_secp384r1_uint1 arg1, uint64_t arg2,
+                                       uint64_t arg3) {
+    fiat_secp384r1_uint1 x1 = (!(!arg1));
+    uint64_t x2 =
+        ((fiat_secp384r1_int1)(0x0 - x1) & UINT64_C(0xffffffffffffffff));
+    uint64_t x3 = ((x2 & arg3) | ((~x2) & arg2));
+    *out1 = x3;
+}
+
+/*
+ * The function fiat_secp384r1_mul multiplies two field elements in the
+ * Montgomery domain. Preconditions: 0 ≤ eval arg1 < m 0 ≤ eval arg2 < m
+ * Postconditions:
+ *   eval (from_montgomery out1) mod m = (eval (from_montgomery arg1) * eval
+ * (from_montgomery arg2)) mod m 0 ≤ eval out1 < m
+ *
+ * Input Bounds:
+ *   arg1: [[0x0 ~> 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~>
+ * 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~>
+ * 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff]] arg2: [[0x0 ~>
+ * 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~>
+ * 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~>
+ * 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff]] Output Bounds: out1: [[0x0
+ * ~> 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~>
+ * 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~>
+ * 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff]]
+ */
+static void fiat_secp384r1_mul(uint64_t out1[6], const uint64_t arg1[6],
+                               const uint64_t arg2[6]) {
+    uint64_t x1 = (arg1[1]);
+    uint64_t x2 = (arg1[2]);
+    uint64_t x3 = (arg1[3]);
+    uint64_t x4 = (arg1[4]);
+    uint64_t x5 = (arg1[5]);
+    uint64_t x6 = (arg1[0]);
+    uint64_t x7;
+    uint64_t x8;
+    fiat_secp384r1_mulx_u64(&x7, &x8, x6, (arg2[5]));
+    uint64_t x9;
```

Comment:
Do we have enough control on fiat to avoid it generating code that mixes variable declarations and code?

Reply:
> Do we have enough control on fiat to avoid it generating code that mixes variable declarations and code?

Fiat lets you control stuff like `static` and selectively emitting functions, but not on this granularity.

I'll file an issue with them.

Reply:
Fixed in 5fc91a34cb069928b0ebf5e08e4ee155980e190f with quick action by @JasonGross (thanks!)